test/config_tests.ml

```diff
@@ -0,0 +1,124 @@
+       | _ -> false)
+
+let conf_map = Alcotest.testable
+    Openvpn.Config.pp Openvpn.Config.(equal eq)
+
+let parse_noextern conf =
+  Openvpn.Config.parse_client ~string_of_file:(fun path ->
+      Rresult.R.error_msgf
+        "this test suite does not read external files, \
+         but a config asked for: %S" path) conf
+
+
+let ok_minimal_client () =
+  (* verify that we can parse a minimal good config. *)
+  let basic =
+    {|tls-client
+    auth-user-pass [inline]
+    <auth-user-pass>
+testuser
+testpass
+</auth-user-pass>
+remote 10.0.0.1|} in
+  Alcotest.(check (result pass pmsg)) "basic conf works"
+    (Ok Openvpn.Config.empty)
+    (parse_noextern basic)
+
+let auth_user_pass_trailing_whitespace () =
+  (* Seems to me that the OpenVPN upstream accepts this and we don't.
+     It should also be tested if the upstream version strips prefixed/trailing
+     whitespace from the user/pass/end-block lines. TODO.
+  *)
+  let common =
+    "client\n"
+    ^ "remote 127.0.0.1\n"
+    ^ "auth-user-pass [inline]\n"
+    ^ "<auth-user-pass>\n"
+    ^ "testuser\n"
+    ^ "testpass\n" in
+  let expected = common ^ "</auth-user-pass>" |> parse_noextern in
+  let with_trailing = common ^ "\n</auth-user-pass>" |> parse_noextern in
+  Alcotest.(check (result_p_err conf_map))
+    "accept trailing whitespace in <auth-user-pass> blocks"
+    expected with_trailing
+
+let rport_precedence () =
+  (* NOTE: at the moment this is expected to fail because we do not implement
+     the rport directive correctly. TODO *)
+  (* see https://github.com/roburio/openvpn/pull/12#issuecomment-581449319 *)
+  let sample =
+    {|
+    tls-client
+    auth-user-pass [inline]
+    <auth-user-pass>
+testuser
+testpass
+</auth-user-pass>
+
+    remote 10.0.42.5
+    remote 10.0.42.3 1194
+    rport 1234
+    remote 10.0.42.4
+|} in
+  let open Openvpn.Config in
+  let sample = parse_client
+      ~string_of_file:(fun _ -> Rresult.R.error_msg "oops")
+      sample
+  in
+  let expected =
+        {|
+    tls-client
+    auth-user-pass [inline]
+    <auth-user-pass>
+testuser
+testpass
+</auth-user-pass>
+
+    remote 10.0.42.5 1234
+    remote 10.0.42.3 1194
+    rport 1234
+    remote 10.0.42.4 1234
+|} |> parse_client ~string_of_file:(fun _ -> Rresult.R.error_msg "")
+      |> Rresult.R.get_ok
+  in
+  Alcotest.check (result_p_err conf_map)
+    "rport doesn't override explicits that coincide with the default"
+    (Ok expected)
+    (sample) ;
+  let _ip1 = Ipaddr.of_string_exn "10.0.42.3" in
+  let _ip2 = Ipaddr.of_string_exn "10.0.42.4" in
+  let _ip3 = Ipaddr.of_string_exn "10.0.42.5" in
+  () (* TODO check that the ports and remotes also match the written *)
+
+let crowbar_fuzz_config () =
+  AflPersistent.run (fun () ->
+  Crowbar.add_test ~name:"Fuzzing doesn't crash Config.parse_client"
+    [Crowbar.bytes] (fun s ->
+        try Crowbar.check (ignore @@ parse_noextern s ; true)
+        with _ -> Crowbar.bad_test ()
+      )
+    )
```

Comment:
I'd recommend to put the unit/regression tests of configuration in a separate binary than the crowbar fuzzing (due to dependencies -- it's nice to run the unit tests on each commit, while fuzzing could be only on drastic changes to the configuration parser).

The fuzzing part atm checks that `parse_noextern` does not crash (IIUC) -- we may as well generate `Conf_map.t` values, serialise (generating a string) and parse that (getting another Conf_map.t) and check that (a) the parse succeeds, (b) the two Conf_map.t are equal. this would catch things like typos in serialiser / parser... (maybe out of scope for this PR)

Reply:
Crowbar runs pretty fast for me, but if you're worried about the dependency, sure. I already have a separate file for `afl-fuzz` that I should also commit.

Good test to have, but yes, I think in a different PR, maybe using the Crowbar+`deriving` tooling.

Reply:
ok, I agree that crowbar + deriving would be nice there. can we move the afl-persistent / crowbar dependent test to `config_afl.ml`? I usually develop in switches that are not based on afl, and as mentioned above would appreciate to run the unit tests nevertheless during development.

Reply:
Yes, will do that :)